tests/cpp/test_host_ir_integration.cpp

```diff
@@ -177,6 +184,86 @@ TEST_F(HostIrIntegrationTest, Deallocate) {
   EXPECT_EQ(memoryAllocated(device_index), 0);
 }
 
+TEST_F(HostIrIntegrationTest, InsertDeallocations) {
+  c10::DeviceIndex device_index = 0;
+  resetPeakMemoryStats(device_index);
+  at::cuda::clearCublasWorkspaces();
+  nvfuser::releaseZeroedMemory();
+  ASSERT_EQ(memoryAllocated(device_index), 0)
+      << "Previous tests leaked memory.";
+
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  std::vector<int64_t> input_shape{32, 32};
+  auto in = TensorViewBuilder()
+                .ndims(input_shape.size())
+                .dtype(DataType::Double)
+                .build();
+
+  fusion->addInput(in);
+  TensorView* t0 = add(in, in);
+  auto t1 = segment_set(t0);
+  TensorView* t2 = add(t1, t1);
+  auto t3 = segment_set(t2);
+  TensorView* t4 = add(t3, t1);
+  TensorView* out = add(t4, t4);
+  fusion->addOutput(out);
+
+  FusionExecutorCache executor_cache(std::move(fusion));
+
+  at::Tensor in_tensor = at::randn(
+      input_shape, at::dtype(at::kDouble).device(at::kCUDA, device_index));
+  auto out_tensors = executor_cache.runFusionWithInputs({in_tensor});
+
+  const int64_t max_memory_allocated = maxMemoryAllocated(device_index);
+
+  FusionKernelRuntime* runtime = executor_cache.getMostRecentKernelRuntime();
+  EXPECT_EQ(runtime->getHostIrEvaluator().canRun(), "");
+  auto hicExprs =
+      runtime->getHostIrEvaluator().getHostIrContainer().topLevelExprs();
+
+  EXPECT_THAT(
+      hicExprs,
+      Contains(Property(&Expr::isA<Deallocate>, IsTrue()))
+          .Times(testing::Eq(3)))
+      << "host ir container should have 3 deallocate ops";
+
+  testValidate(
+      executor_cache.fusion(),
+      out_tensors,
+      {in_tensor},
+      __LINE__,
+      __FILE__,
+      "");
+
+  if (c10::utils::check_env("PYTORCH_NO_CUDA_MEMORY_CACHING")) {
+    GTEST_SKIP() << "Skipped because PYTORCH_NO_CUDA_MEMORY_CACHING is on. "
+                    "This usually happens when running with compute-sanitizer. "
+                    "maxMemoryAllocated can only collect peak memory "
+                    "from a caching allocator.";
+  }
+
+  // At any given time a max of four 32x32 tensors are allocated. Here is the
```

Comment:
Thanks for the comment! 

IIUC, the test doesn't really show the end benefit of deallocation because the peak memory would be the same without deallocation? I'd consider making t4 a unary (e.g. sin) of t3 so `t1` can be deallocated earlier.